### PR TITLE
[MSFT] Re-work ExportQuartusTcl to export into verbatim nodes.

### PR DESCRIPTION
**Pull request by mikeurbach**
Start towards https://github.com/llvm/circt/issues/2068. I think this does the important part at the C++ level the intended way, but there are some TODOs I want to look into related to controlling where the Tcl file actually goes. I will put back the `msft.export_tcl` API after that.

### CODE REVIEW

**Comment by mikeurbach** [comment]
@teqdruid I think this is good for review as a starting point. I will have two follow up PRs, that can come after this: one to pull this pass up into the `lower-msft-to-hw` level, and another to switch from emitting verilog into a buffer to emitting into files. I think those can come separately for the sake of incremental PRs.

File: integration_test/Bindings/Python/dialects/msft.py [hunk 1]
```diff
@@ -85,13 +85,6 @@
   assert num_failed == 1
   # ERR: error: 'hw.instance' op Could not apply placement #msft.physloc<M20K, 2, 6, 1>. Position already occupied by hw.instance "ext1" @MyExternMod
 
-  # CHECK-LABEL: === tcl ===
-  print("=== tcl ===")
-
-  # CHECK: proc top_config { parent } {
-  # CHECK:   set_location_assignment M20K_X2_Y6_N1 -to $parent|inst1|ext1|ext1|foo_subpath
```
Comment (teqdruid) [inline integration_test/Bindings/Python/dialects/msft.py:92]: Can you keep these filecheck lines?

File: lib/Dialect/MSFT/ExportQuartusTcl.cpp [hunk 2]
```diff
@@ -37,18 +41,32 @@ struct TclOutputState {
...
-void emitPath(TclOutputState &s, RootedInstancePathAttr path) {
-  for (auto part : path.getPath())
-    s.os << part.getValue() << '|';
+void emitPath(TclOutputState &s, RootedInstancePathAttr path,
+              SymbolCache &symCache) {
+  for (auto part : path.getPath()) {
+    auto inst = cast<msft::InstanceOp>(symCache.getDefinition(part));
+    assert(inst && "path instance must be in symbol cache");
+    auto mod = inst->getParentOfType<MSFTModuleOp>();
+    assert(mod && "path instance must have MSFT module parent");
```
Comment (teqdruid) [inline lib/Dialect/MSFT/ExportQuartusTcl.cpp:55]: I'd rather not have this restriction. (Does it existing in the tblgen def?) Since it looks like you only need the symbol, could you use the MLIR symbol stuff instead?

File: lib/Dialect/MSFT/ExportQuartusTcl.cpp [hunk 3]
```diff
@@ -37,18 +41,32 @@ struct TclOutputState {
...
+
+    s.os << "{{" << s.symbolRefs.size() << "}}" << '|';
```
Comment (teqdruid) [inline lib/Dialect/MSFT/ExportQuartusTcl.cpp:57]: Took me a minute to figure out what's going on in the rest of the lines. Maybe add a code comment?

File: lib/Dialect/MSFT/ExportQuartusTcl.cpp [hunk 4]
```diff
@@ -73,34 +91,71 @@ static void emit(TclOutputState &s, PlacementDB::PlacedInstance inst,
...
+  std::string s;
+  llvm::raw_string_ostream os(s);
   TclOutputState state(os);
   PlacementDB db(hwMod);
   size_t failures = db.addDesignPlacements();
   if (failures != 0)
     return hwMod->emitError("Could not place ") << failures << " instances";
 
+  // Don't emit empty procs if the database is empty.
+  if (db.empty())
```
Comment (teqdruid) [inline lib/Dialect/MSFT/ExportQuartusTcl.cpp:136]: I don't think we want to skip emitting a proc which the user has requested be generated. Other tcl may depend on it being there, even if it is empty.

File: lib/Dialect/MSFT/MSFTPasses.cpp [hunk 5]
```diff
@@ -146,6 +173,14 @@ void LowerToHWPass::runOnOperation() {
   auto top = getOperation();
   auto *ctxt = &getContext();
 
+  // Traverse MSFT location attributes and export the required Tcl into
+  // templated `sv::VerbatimOp`s with symbolic references to the instance paths.
+  hw::SymbolCache symCache;
+  populateSymbolCache(top, symCache);
+  for (auto hwmod : top.getBody()->getOps<msft::MSFTModuleOp>())
```
Comment (teqdruid) [inline lib/Dialect/MSFT/MSFTPasses.cpp:180]: Rather than relying on the devdb being empty, plz use a pass option containing a list of 'tops'. This could be a separate PR.

**Review by teqdruid — APPROVED** [review]
Looks great! Thanks for doing this. Just a few changes at the edges then I think we're good to go!

File: lib/Dialect/MSFT/MSFTPasses.cpp (hunk 5 above)
Comment (mikeurbach) [reply lib/Dialect/MSFT/MSFTPasses.cpp:180]: Got it, will do.

File: lib/Dialect/MSFT/ExportQuartusTcl.cpp (hunk 4 above)
Comment (mikeurbach) [reply lib/Dialect/MSFT/ExportQuartusTcl.cpp:136]: Ah, makes sense. I will roll it into the change to accept a list of top modules to generate Tcl for.

File: lib/Dialect/MSFT/ExportQuartusTcl.cpp (hunk 2 above)
Comment (mikeurbach) [reply lib/Dialect/MSFT/ExportQuartusTcl.cpp:55]: This was left from old code that looked up either a HWModuleOp or MSFTModuleOp, and had to confirm it found something. It is something we can guarantee in the tblgen def, and I can indeed just use the module symbol.

Comment (teqdruid) [reply lib/Dialect/MSFT/ExportQuartusTcl.cpp:55]: I'm debating whether or not we want to be restrictive about where msft.instances can live. Anywhere or just in msft.modules? Should we start with restrictive then be more open in the future if necessary?

File: integration_test/Bindings/Python/dialects/msft.py (hunk 1 above)
Comment (mikeurbach) [reply integration_test/Bindings/Python/dialects/msft.py:92]: These depend on the now-removed `msft.export_tcl` API. I can run the lower-msft-to-tcl pass and check for the output that way.

File: include/circt/Dialect/MSFT/DeviceDB.h [hunk 6]
```diff
@@ -102,6 +102,9 @@ class PlacementDB {
                           std::make_tuple(-1, -1, -1, -1),
                       Optional<PrimitiveType> primType = {});
 
+  /// Helper function to check if the database is empty.
+  bool empty();
```
[the remark this answers is not in the document]
Comment (mikeurbach) [reply include/circt/Dialect/MSFT/DeviceDB.h:106]: Will be removed.

File: integration_test/Bindings/Python/dialects/msft.py (hunk 1 above)
Comment (teqdruid) [reply integration_test/Bindings/Python/dialects/msft.py:92]: Please do

File: lib/Dialect/MSFT/ExportQuartusTcl.cpp (hunk 2 above)
Comment (mikeurbach) [reply lib/Dialect/MSFT/ExportQuartusTcl.cpp:55]: Right now they have to support MSFTModule or HWModule parents to support partial lowering during lower-msft-to-hw (MSFTModule is lowered to HWModule first, then the MSFTInstance is lowered). Tablegen already requires the parent to be either MSFTModule or HWModule and nothing else. I think keeping that restriction for now makes sense.

Comment (teqdruid) [reply lib/Dialect/MSFT/ExportQuartusTcl.cpp:55]: Yep, agreed.

File: integration_test/Bindings/Python/dialects/msft.py (hunk 1 above)
Comment (mikeurbach) [reply integration_test/Bindings/Python/dialects/msft.py:92]: So now that I'm looking at this, this test is building a mix of HW and MSFT dialect ops, which isn't compatible with the contract lower-msft-to-tcl expects. I can re-work this test to use only MSFT ops in order to test the Tcl output, but that is a bigger change. Or I can just leave off the Tcl output checks, as it is now.

Comment (mikeurbach) [reply integration_test/Bindings/Python/dialects/msft.py:92]: Ok, I've untangled the issue and can add back these checks, they will just have to go at the end of the file, since we modify the IR with lower-msft-to-hw before we can export tcl.

File: lib/Dialect/MSFT/MSFTPasses.cpp (hunk 5 above)
Comment (mikeurbach) [reply lib/Dialect/MSFT/MSFTPasses.cpp]: @teqdruid I implemented this on the current PR, since I didn't want to land the new `empty` method just to remove it later. From PyCDE, we pass use the list maintained in `self.symbols`, is that what you had in mind?

File: frontends/PyCDE/src/pycde/system.py [hunk 7]
```diff
@@ -35,26 +37,36 @@ class System:
...
+    lower-msft-to-hw{{tops={tops} verilog-file={verilog_file} tcl-file={tcl_file}}},
+    lower-seq-to-sv,hw.module(prettify-verilog),hw.module(hw-cleanup)
+  """
 
-  def __init__(self, modules, primdb: PrimitiveDB = None):
+  def __init__(self,
+               modules,
+               primdb: PrimitiveDB = None,
+               name: str = "PyCDESystem",
+               output_directory: str = None):
```
Comment (mikeurbach) [inline frontends/PyCDE/src/pycde/system.py:54]: @teqdruid ended up adding this output_directory for the tests. For the tests to be stable, we want to always write the output to the `%t` directory lit provides, and ensure `%t` is cleaned up of previous runs before running the tests. Since `%t` isn't just os.getcwd() when the tests run, we need to pass it explicitly. It still defaults to os.getcwd() for normal use.

**Comment by mikeurbach** [comment]
I think this is good to go. Let me know if there is anything else to adjust and I can do it in a follow up.
